### PR TITLE
Implement OIDC flow & node server

### PR DESCRIPTION
**Pull request by dulithsenanayake**
This PR includes OIDC flow implementation & node server implementation

### CODE REVIEW

File: portals/new_admin/portals/bin/www [hunk 1]
```diff
@@ -0,0 +1,95 @@
+#!/usr/bin/env node
+
+/**
+ * Module dependencies.
+ */
+
+var app = require('../app');
+var debug = require('debug')('apps:server');
+// var http = require('http');
```
Comment (senthuran16) [inline portals/new_admin/portals/bin/www:9]: Shall we remove this commented line if we don't need this?

File: portals/new_admin/portals/client/src/crypto.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+//     return text;
+// }
+
+// export async function generateCodeChallenge(codeVerifier) {
+//     var digest = await crypto.subtle.digest("SHA-256",
+//         new TextEncoder().encode(codeVerifier));
+
+//     return btoa(String.fromCharCode(...new Uint8Array(digest)))
+//         .replace(/=/g, '').replace(/\+/g, '-').replace(/\//g, '_')
+// }
```
Comment (senthuran16) [inline portals/new_admin/portals/client/src/crypto.ts:53]: Shall we remove this commented code?

File: portals/new_admin/portals/client/src/models/oidc-request-params.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+/**
+ * OIDC request parameters.
+ */
+ export interface OIDCRequestParamsInterface {
+    clientId: string;
+    // redirectUri: string;
```
Comment (senthuran16) [inline portals/new_admin/portals/client/src/models/oidc-request-params.ts:6]: Shall we remove this commented line if not needed?

File: portals/new_admin/portals/client/src/sign-in.ts [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+//                 redirect_uri: "https://localhost:4000/#/publisher/token",
+//                 response_type: "code",
+//                 scope: "openid",
+//                 state: "RlZyVjlqYUpHTzltWC42c2FNRDRJT1JPfk1+TUFEa0RLb04yZldwYkpxVA==",
+//                 code_challenge: codeChallenge,
+//                 code_challenge_method: "S256"
+//             });
+//             window.location = "https://dev-kw-oeodk.us.auth0.com/authorize?" + args;
+//         })
+// }
```
Comment (senthuran16) [inline portals/new_admin/portals/client/src/sign-in.ts:107]: Shall we remove this commented code if not needed?

File: portals/new_admin/portals/webpack.config.js [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+const path = require('path');
+
+module.exports = (env) => {
+  // const env = dotenv.config().parsed;
+
+  // // reduce it to a nice object, the same as before
+  // const envKeys = Object.keys(env).reduce((prev, next) => {
+  //   prev[`process.env.${next}`] = JSON.stringify(env[next]);
+  //   return prev;
+  // }, {});
```
Comment (senthuran16) [inline portals/new_admin/portals/webpack.config.js:11]: Shall we remove this commented code if not needed?

File: portals/new_admin/portals/client/src/Apis.js [hunk 6]
```diff
@@ -0,0 +1,76 @@
+import React, { useEffect, useState } from 'react'
```
Comment (ashera96) [inline portals/new_admin/portals/client/src/Apis.js:1]: Don't we need the license headers for these files?

File: portals/new_admin/portals/tsconfig.json [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+        "sourceMap": true,
+        "noImplicitAny": true,
+        "module": "es6",
+        "target": "es5",
+        "jsx": "react",
+        "allowJs": true,
+        "moduleResolution": "node",
+        "allowSyntheticDefaultImports": true
+    }
+}
```
Comment (ashera96) [inline portals/new_admin/portals/tsconfig.json:13]: Shall we fix the EOF newline missing issue? Please fix it in other places too

File: portals/new_admin/portals/routes/users.js [hunk 8]
```diff
@@ -0,0 +1,9 @@
+var express = require('express');
+var router = express.Router();
+
+/* GET users listing. */
+router.get('/', function(req, res, next) {
+  res.send('respond with a resourceeezzz333');
```
Comment (ashera96) [inline portals/new_admin/portals/routes/users.js:6]: Why have we used `resourceeezzz333` here? Was this done as a demo?

File: portals/new_admin/portals/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,10 @@
+FROM node:16-alpine
+WORKDIR /app
+COPY . .
+RUN npm i
+RUN ls
+RUN echo "$PWD"
+RUN npm install
+RUN npm run react-prod
+EXPOSE 4000
+CMD ["npm", "run", "start"]
```
Comment (wasuradananjith) [inline portals/new_admin/portals/Dockerfile:10]: Shall we add an EOF new line here?

File: portals/new_admin/portals/client/dist/index.html [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+    <title>React Webpack App</title>
+</head>
+
+<body>
+    <div id="root"></div>
+    <script src="bundle.js"></script>
+</body>
+
+</html>
```
Comment (wasuradananjith) [inline portals/new_admin/portals/client/dist/index.html:15]: Shall we add an EOF new line here as well?

File: portals/new_admin/portals/client/public/index.html [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+<html>
+  <head>
+    <meta charset="utf-8"/>
+    <title>WSO2 API Manager</title>
+    <script type="text/javascript" src="/Settings.js"></script>
+  <script defer src="/bundle.js"></script></head>
+  <body>
+    <div id="root"></div>
+  </body>
+</html>
```
Comment (wasuradananjith) [inline portals/new_admin/portals/client/public/index.html:11]: Shall we add an EOF new line here as well?

File: portals/new_admin/portals/client/src/App.css [hunk 12]
```diff
@@ -0,0 +1,469 @@
...
+}
+
+@keyframes h1 {
+  from {
+    transform: translateY(180%);
+  }
+  to {
+    transform: translateY(0);
+  }
+}
```
Comment (wasuradananjith) [inline portals/new_admin/portals/client/src/App.css:469]: Shall we add an EOF new line here as well?

File: portals/new_admin/portals/client/src/App.js [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+function NoMatch() {
+  return (
+    <div>
+      <h2>Nothing to see here!</h2>
+      <p>
+        <Link to="/">Go to the home page</Link>
+      </p>
+    </div>
+  );
+}
```
Comment (wasuradananjith) [inline portals/new_admin/portals/client/src/App.js:35]: Shall we add an EOF new line here as well?

File: portals/new_admin/portals/client/src/LoginButton.tsx [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+  const requestParams: OIDCRequestParamsInterface = {
+    clientId: 'FbCSH23HybQMV9UlXJfeKHogAEHojzCO',
+    scope: 'openid',
+    state: 'RlZyVjlqYUpHTzltWC42c2FNRDRJT1JPfk1+TUFEa0RLb04yZldwYkpxVA=='
+  };
+
+  return <button onClick={() => sendAuthorizationRequest(requestParams)}>Log In</button>;
+};
+
+export default LoginButton;
```
Comment (wasuradananjith) [inline portals/new_admin/portals/client/src/LoginButton.tsx:15]: Shall we add an EOF new line here as well?

File: portals/new_admin/portals/client/src/LogoutButton.js [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+  const { logout } = useAuth0();
+
+  return (
+    <button onClick={() => logout({ returnTo: window.location.origin })}>
+      Log Out
+    </button>
+  );
+};
+
+export default LogoutButton;
```
Comment (wasuradananjith) [inline portals/new_admin/portals/client/src/LogoutButton.js:14]: Shall we add an EOF new line here as well?

File: portals/new_admin/portals/client/src/Profile.js [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+      <div>
+        <img src={user.picture} alt={user.name} />
+        <h2>{user.name}</h2>
+        <p>{user.email}</p>
+      </div>
+    )
+  );
+};
+
+export default Profile;
```
Comment (wasuradananjith) [inline portals/new_admin/portals/client/src/Profile.js:22]: Shall we add an EOF new line here as well?

File: portals/new_admin/portals/client/src/TokenProcessor.tsx [hunk 17]
```diff
@@ -0,0 +1,16 @@
...
+  const requestParams: OIDCRequestParamsInterface = {
+    clientId: 'FbCSH23HybQMV9UlXJfeKHogAEHojzCO',
+    scope: 'openid',
+    state: 'RlZyVjlqYUpHTzltWC42c2FNRDRJT1JPfk1+TUFEa0RLb04yZldwYkpxVA=='
+  };
+  sendTokenRequest(requestParams);
+  return <></>;
+}
+
+export default TokenProcessor;
```
Comment (wasuradananjith) [inline portals/new_admin/portals/client/src/TokenProcessor.tsx:16]: Shall we add an EOF new line here as well?

File: portals/new_admin/portals/client/src/components/App.jsx [hunk 18]
```diff
@@ -0,0 +1,17 @@
...
+
+  render() {
+    return (
+      <h1>Hello World0!</h1>
+    );
+  }
+}
+
+
+export default App;
```
Comment (wasuradananjith) [inline portals/new_admin/portals/client/src/components/App.jsx:17]: Shall we add an EOF new line here as well?

File: portals/new_admin/portals/client/src/constants/token.ts [hunk 19]
```diff
@@ -0,0 +1 @@
+export const AUTHORIZATION_CODE = "code";
```
Comment (wasuradananjith) [inline portals/new_admin/portals/client/src/constants/token.ts:1]: Shall we add an EOF new line here as well?

File: portals/new_admin/portals/client/src/crypto.ts (hunk 2 above)
Comment (wasuradananjith) [inline portals/new_admin/portals/client/src/crypto.ts:53]: Shall we add an EOF new line here as well?

File: portals/new_admin/portals/client/src/index.html [hunk 20]
```diff
@@ -0,0 +1,11 @@
...
+<html>
+  <head>
+    <meta charset="utf-8"/>
+    <title><%= htmlWebpackPlugin.options.title %></title>
+    <script type="text/javascript" src="/Settings.js"></script>
+  </head>
+  <body>
+    <div id="root"></div>
+  </body>
+</html>
```
Comment (wasuradananjith) [inline portals/new_admin/portals/client/src/index.html:11]: Shall we add an EOF new line here as well?

File: portals/new_admin/portals/client/src/index.jsx [hunk 21]
```diff
@@ -0,0 +1,16 @@
...
+import Settings from 'Settings';
+
+const root = ReactDOM.createRoot(document.getElementById('root'));
+root.render(
+  <React.StrictMode>
+    <BrowserRouter >
+    <App />
+  </BrowserRouter>
+  </React.StrictMode>
+);
```
Comment (wasuradananjith) [inline portals/new_admin/portals/client/src/index.jsx:16]: Shall we add an EOF new line here as well?

File: portals/new_admin/portals/client/src/models/oidc-request-params.ts [hunk 22]
```diff
@@ -0,0 +1,9 @@
...
+    scope: string;
+    state: string
+}
```
Comment (wasuradananjith) [inline portals/new_admin/portals/client/src/models/oidc-request-params.ts:9]: Shall we add an EOF new line here as well?

File: portals/new_admin/portals/client/src/models/token-response.ts [hunk 23]
```diff
@@ -0,0 +1,10 @@
+/**
+ * Interface of the OAuth2/OIDC tokens.
+ */
+ export interface TokenResponseInterface {
+    accessToken: string;
+    idToken: string;
+    expiresIn: string;
+    scope: string;
+    tokenType: string;
+}
```
Comment (wasuradananjith) [inline portals/new_admin/portals/client/src/models/token-response.ts:10]: Shall we add an EOF new line here as well?

File: portals/new_admin/portals/tsconfig.json (hunk 7 above)
Comment (wasuradananjith) [reply portals/new_admin/portals/tsconfig.json:13]: +1

File: portals/new_admin/portals/webpack.config.js [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+    return {
+      ...devConfig
+    }
+  } else {
+    return {
+      mode: 'development',
+      ...devConfig
+    }
+  }
+};
```
Comment (wasuradananjith) [inline portals/new_admin/portals/webpack.config.js:72]: Shall we add an EOF new line here as well?

File: portals/new_admin/portals/client/src/crypto.ts (hunk 2 above)
Comment (dulithsenanayake) [reply portals/new_admin/portals/client/src/crypto.ts:53]: These commented lines are not there in the latest file.

File: portals/new_admin/portals/client/src/sign-in.ts (hunk 4 above)
Comment (dulithsenanayake) [reply portals/new_admin/portals/client/src/sign-in.ts:107]: These commented lines are not there in the latest file.

File: portals/new_admin/portals/routes/users.js (hunk 8 above)
Comment (dulithsenanayake) [reply portals/new_admin/portals/routes/users.js:6]: Yes, here this was done as a demo.

File: portals/new_admin/portals/client/dist/index.html (hunk 10 above)
Comment (dulithsenanayake) [reply portals/new_admin/portals/client/dist/index.html:15]: Currently, there is no such file.

File: portals/new_admin/portals/client/public/index.html (hunk 11 above)
Comment (dulithsenanayake) [reply portals/new_admin/portals/client/public/index.html:11]: Currently, there is no such file.

File: portals/new_admin/portals/client/src/App.css (hunk 12 above)
Comment (dulithsenanayake) [reply portals/new_admin/portals/client/src/App.css:469]: Currently, there is no such file.

File: portals/new_admin/portals/client/src/Profile.js (hunk 16 above)
Comment (dulithsenanayake) [reply portals/new_admin/portals/client/src/Profile.js:22]: Currently, there is no such file.

File: portals/new_admin/portals/client/src/TokenProcessor.tsx (hunk 17 above)
Comment (dulithsenanayake) [reply portals/new_admin/portals/client/src/TokenProcessor.tsx:16]: Currently, there is no such file.

File: portals/new_admin/portals/client/src/components/App.jsx (hunk 18 above)
Comment (dulithsenanayake) [reply portals/new_admin/portals/client/src/components/App.jsx:17]: Currently, there is no such file.

File: portals/new_admin/portals/client/src/constants/token.ts (hunk 19 above)
Comment (dulithsenanayake) [reply portals/new_admin/portals/client/src/constants/token.ts:1]: This issue is fixed in the latest file.

File: portals/new_admin/portals/client/src/index.html (hunk 20 above)
Comment (dulithsenanayake) [reply portals/new_admin/portals/client/src/index.html:11]: Currently, there is no such file.

**Pull request merged**